geyser-plugin-manager/src/accounts_update_notifier.rs

```diff
@@ -74,6 +74,8 @@ impl AccountsUpdateNotifierInterface for AccountsUpdateNotifierImpl {
         }
 
         for plugin in plugin_manager.plugins.iter_mut() {
+            let mut plugin = plugin.write().unwrap();
```

Comment:
Do not take write lock, it will hurt concurrency. Take the read lock

Reply:
All of the relevant methods in `GeyserPlugin` take a `&mut self`, for example, here is the `update_account` signature:

```rust
    #[allow(unused_variables)]
    fn update_account(
        &mut self,
        account: ReplicaAccountInfoVersions,
        slot: u64,
        is_startup: bool,
    ) -> Result<()> {
        Ok(())
    }
   ```
   
   https://github.com/solana-labs/solana/blob/4639919c9b087b68bc24a736fbed26105447b011/geyser-plugin-interface/src/geyser_plugin_interface.rs#L265-L277

Reply:
Right -- we need to think carefully. mut meaning is mutable -- it does not mean it is mutually exclusive -- concurrent operations can happen to a mut object and it is happening.

Reply:
The code isn't going to compile with a read lock. There can exist only one mutable reference for a type at any given time.

Reply:
But I think what you might be going for or thinking about is that we've already (just) taken a lock on the plugin manager so we shouldn't also need a lock on the plugin.

Reply:
Good point. I need to review how lock is being done overall in geyser. We might be hurting concurrency unnecessarily.

Reply:
I don't think there's a way to make concurrent calls into the workhorse functions of a `GeyserPlugin` without breaking changes. You would have to require that the plugin take in `&self` on all methods and require plugin writers to use interior mutability to update plugin state. Maybe a `GeyserPluginV2` can come online down the road. With the current trait interface, locks are needed.

Taking a lock twice is mildly expensive, but not prohibitively so. 1 million double locks is 20-40ms on whatever hardware the playground uses.

https://play.rust-lang.org/?version=stable&mode=release&edition=2021&gist=4fde2ff5764268dae083e2a53ab364e6

Reply:
I also don't know if there are any assumptions baked into the geyser plugin, such as account state changes being received in consensus order. Adding a concurrent-friendly interface to that may break those assumptions.

Reply:
Sound right. Keep this code for now. I will address the overall locking separately. Yes, I might have to break the interface, but it is necessary. We have found validator cannot keep up with the network with plugin loaded.

Reply:
We've been seeing our nodes (RPC) fall behind as well. They have plenty of cpu/mem/bandwidth capacity available, so it's been very confusing to debug. This sequential write lock explanation makes sense. 

@lijunwangs would love to help in any way we can to resolve this.

Reply:
@lijunwangs Would love if you could offer any feedback on this one. I was able to get a `solana-test-validator` running locally with a plugin and didn't notice any issues. I'm going to proceed by cherry-picking back onto v1.14 and deploying a canary to devnet and testnet. 

https://github.com/solana-labs/solana/pull/30380

Reply:
I think we can table this lock issue now for this PR. I will address it separately. My major concern is how the system detect the need to reload. I would like to implement via the admin RPC. So the flow likes the following

User ---> Admin RPC --> Send (load/reload) request --> Plugin Manager --> Load/Reload a plugin.

This way we do not have to busily scan for reload which is rare and once it is requested we can quickly handle it. In addition the same mechanism can be used to load a new plugin.

Reply:
Yea that sounds great. Are there any admin RPC methods I could use a reference? Can I get a permalink?

Reply:
cc. @t-nelson @CriesofCarrots